### PR TITLE
[design-v2] Add NEP9 QR code creation to receive screen

### PR DESCRIPTION
**Pull request by nickfujita**
**What problem does this PR solve?**
The current receive screen shows the QR code for the users address. Now that [NEP9](https://github.com/neo-project/proposals/pull/25) is accepted, adding across the apps.

**How did you solve this problem?**
Added [dynamically generated QR codes](https://o3.network/nep9/) with logos for NEO/GAS/NEP5

**How did you make sure your solution works?**
- Navigate to receive screen
- Presented with base QR code with just address, in NEP9 URI format
image

- Add in parameters for the asset to receive, or the amount
image

### CODE REVIEW

File: app/components/Receive/ReceivePanel/ReceivePanel.jsx [hunk 1]
```diff
@@ -14,26 +18,48 @@ type Props = {
...
   render () {
     const { className, address } = this.props
+    const { asset, amount } = this.state
+    const symbols = ['-', 'NEO', 'GAS', ...Object.keys(TOKENS)]
 
     return (
       <Panel className={classNames(styles.receivePanel, className)} renderHeader={this.renderHeader}>
         <div className={styles.qrcode}>
-          <canvas ref={this.registerRef} />
+          <img ref={this.registerRef} />
```
Comment (evgenyboxer) [inline app/components/Receive/ReceivePanel/ReceivePanel.jsx]: I think its ok to do it inline `ref={(el: HTMLImageElement) => { this.image = el })`

File: app/components/Receive/ReceivePanel/ReceivePanel.jsx [hunk 2]
```diff
@@ -44,7 +70,48 @@ export default class ReceivePanel extends React.Component<Props> {
...
+    const qrCode = new NeoQR(options, 250)
+    qrCode.toDataURL()
+      .then(imgData => {
+        if (this.image) {
+          this.image.src = imgData
+        }
+      })
+  }
+
+  getSymbols = () => {
```
Comment (evgenyboxer) [inline app/components/Receive/ReceivePanel/ReceivePanel.jsx]: is this function needed?

File: app/components/Receive/ReceivePanel/ReceivePanel.jsx [hunk 3]
```diff
@@ -14,26 +18,48 @@ type Props = {
...
+
+  state = {
+    asset: null,
+    amount: 0
   }
 
   render () {
     const { className, address } = this.props
+    const { asset, amount } = this.state
+    const symbols = ['-', 'NEO', 'GAS', ...Object.keys(TOKENS)]
```
Comment (evgenyboxer) [inline app/components/Receive/ReceivePanel/ReceivePanel.jsx]: Could you use `ASSETS.NEO` and `ASSETS.GAS` for NEO/GAS

File: app/components/Receive/ReceivePanel/ReceivePanel.jsx [hunk 4]
```diff
@@ -44,7 +70,48 @@ export default class ReceivePanel extends React.Component<Props> {
...
+  componentDidUpdate (prevProps: Props, prevState: State) {
+    const { address } = this.props
+    const { asset, amount } = this.state
+
+    const assetChanged = prevState.asset !== asset
+    const amountChanged = prevState.amount !== amount
+
+    if (assetChanged || amountChanged) {
+      let convertedAsset
+      if (asset === 'NEO' || asset === 'GAS') {
```
Comment (evgenyboxer) [inline app/components/Receive/ReceivePanel/ReceivePanel.jsx]: Use `[ASSETS.NEO, ASSETS.GAS].includes(asset)` instead

**Comment by evgenyboxer** [comment]
@nickfujita awesome job with NEP9. However, I think the design might be a bit confusing because its not clear what is the purpose of the symbol/amount.

I'll forward this to our designer

File: app/components/Receive/ReceivePanel/ReceivePanel.jsx (hunk 1 above)
Comment (nickfujita) [reply app/components/Receive/ReceivePanel/ReceivePanel.jsx]: updated

File: app/components/Receive/ReceivePanel/ReceivePanel.jsx (hunk 2 above)
Comment (nickfujita) [reply app/components/Receive/ReceivePanel/ReceivePanel.jsx]: nope :)
cleaned up

File: app/components/Receive/ReceivePanel/ReceivePanel.jsx (hunk 3 above)
Comment (nickfujita) [reply app/components/Receive/ReceivePanel/ReceivePanel.jsx]: updated

File: app/components/Receive/ReceivePanel/ReceivePanel.jsx (hunk 4 above)
Comment (nickfujita) [reply app/components/Receive/ReceivePanel/ReceivePanel.jsx]: updated

**Comment by nickfujita** [comment]
Code updates complete. Awaiting feedback on design.

**Comment by 0ben0** [comment]
@evgenyboxer 

Does this flow work?

image

**Comment by nickfujita** [comment]
@0ben0 Is it possible to just keep the QR code image and image URL on the same page as the receive, rather than in a separate modal?

**Comment by 0ben0** [comment]
@nickfujita 

I've used modals throughout the design - so was replicating functionality here. Have you seen the completed designs?

[Neon_WIP_05.pdf](https://github.com/CityOfZion/neon-wallet/files/2128587/Neon_WIP_05.pdf)

If we change it here, does that mean it'll need removing/ redesigning in all other areas of the app?

**Comment by comountainclimber** [comment]
@nickfujita I know it sounds painful but can you cherry-pick this work against design-v2 and rework the UX based on the mocks?

**Comment by comountainclimber** [comment]
@nickfujita are you still willing to take on this work??

**Comment by nickfujita** [comment]
@comountainclimber Sure, I thought we were still waiting on design feedback from you and @evgenyboxer 

If it's been finalized, I can move forward with the updates.

**Comment by nickfujita** [comment]
Looking over the designs, a couple questions came up.
- On the `Amount` field, what is meant to happen when the `Max` button is clicked? Since it is receive, the requested amount could be any amount.
- There appears to be an option to change the address to receive the requested funds. Should this not just be set to the address of the logged in account?
- Is there not a way for the user to just get the QR code for their address without generating a NEP9?
- How do we determine the what assets are shown at the top bar with quantities and prices? What if the wallet has no assets? What is the wallet has more than 4 assets?

**Comment by comountainclimber** [comment]
@0ben0 when you have the opportunity can you address the above items so we can wrap up this work and get into the design-v2 branch

**Comment by 0ben0** [comment]
@nickfujita @comountainclimber 

Sorry for the delayed response. I've addressed Nick's concerns here - Max, please weigh in if required:

1. **On the Amount field, what is meant to happen when the Max button is clicked? Since it is receive, the requested amount could be any amount.**
 _This is an error on my part - please feel free to remove._

2. **There appears to be an option to change the address to receive the requested funds. Should this not just be set to the address of the logged in account?**
 _Please remove the drop-down but keep the label as a flag to the user for the destination wallet_

3. **Is there not a way for the user to just get the QR code for their address without generating a NEP9?** _I don't think so. However, this is above my pay grade - Max?_

4. **How do we determine the what assets are shown at the top bar with quantities and prices? What if the wallet has no assets? What is the wallet has more than 4 assets?**
_My understanding is that this will come from the supported list of NEP5 tokens? The latest designs have an empty state_

> I'm happy to make any changes you need, if it'd be useful? 👍

**Comment by nickfujita** [comment]
Just a last clarification on the amounts at the top. The screenshot below is of the amounts section of this tab. The data can be sourced from the balances of the currently logged in wallet, but what happens in the following cases:

- Wallet is empty
- Wallet has more than 4 tokens
- Wallet has 1-3 tokens
- Should NEO and GAS always be there even if the wallet has none?

image

**Comment by comountainclimber** [comment]
> Wallet is empty
Please reference the mocks below: 
_not currently implemented yet_
https://github.com/CityOfZion/neon-wallet/blob/design-v2/design/_Dark/4.0-Send-Dark%20-%20EMPTY.jpg

> Wallet has more than 4 tokens
This component should scroll horizontally (current implementation)

> Wallet has 1-3 tokens
This component will always center its content

> Should NEO and GAS always be there even if the wallet has none?
Lets stick with the "Nothing to see here" design for now

**Comment by nickfujita** [comment]
@evgenyboxer @comountainclimber Ready for re-review.

**Comment by comountainclimber** [comment]
After pulling down and testing locally - this looks fantastic nice job @nickfujita! Going to leave some comments on code but besides the merge conflicts overall this looks great.

Do you think you have time to implement the "Enter QR Code" feature in the Send UX? The mockups are in /design and the common components have all already been created. Please let me know if you are up for the task

https://github.com/CityOfZion/neon-wallet/blob/design-v2/design/_Dark/4.3-Send-Dark%20QR%20-%20A.jpg

https://github.com/CityOfZion/neon-wallet/blob/design-v2/design/_Dark/4.3-Send-Dark%20QR%20-%20B.jpg

(Please ignore the wallet dropdown as that has been eliminated from the design)

File: app/components/Receive/QRCode/index.jsx [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+
+  state = {
+    asset: null,
+    amount: 0
+  }
+
+  render() {
+    const { className, address } = this.props
+    const { asset, amount } = this.state
+    const symbols = ['-', ASSETS.NEO, ASSETS.GAS, ...Object.keys(TOKENS)]
```
Comment (comountainclimber) [inline app/components/Receive/QRCode/index.jsx]: Do we actually want the '-' here?

File: app/components/Receive/QRCodeForm/index.jsx [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+  state = {
+    asset: undefined,
+    amount: undefined,
+    description: undefined
+  }
+
+  render() {
+    const { className, address, onSubmit } = this.props
+    const { asset, amount, description } = this.state
+    const symbols = ['-', ASSETS.NEO, ASSETS.GAS, ...Object.keys(TOKENS)]
```
Comment (comountainclimber) [inline app/components/Receive/QRCodeForm/index.jsx]: Same as comment above... seems to lead to an invalid option in the select input 
screen shot 2018-09-18 at 7 12 29 pm

File: app/components/Receive/ReceiveExplanation/index.jsx [hunk 7]
```diff
@@ -0,0 +1,35 @@
+// @flow
+import React from 'react'
+import classNames from 'classnames'
+
+import InfoIcon from '../../../assets/icons/info.svg'
+
+import styles from './styles.scss'
+
+type Props = {}
```
Comment (comountainclimber) [inline app/components/Receive/ReceiveExplanation/index.jsx]: I dont think we need this empty type declaration same as below

File: app/components/Receive/ReceiveExplanation/styles.scss [hunk 8]
```diff
@@ -0,0 +1,28 @@
+.receiveExplanation {
+  width: 250px;
```
Comment (comountainclimber) [inline app/components/Receive/ReceiveExplanation/styles.scss]: This doesnt seem to be wide enough...  the container in sketch is 300px can we use that instead

**Comment by nickfujita** [comment]
Feedback addressed. Will tackle the send scanner in separate PR.

**Pull request merged**